### ISSUE TITLE
Pure neovide "copy/paste" support

### ISSUE THREAD

**Issue by j-xella** [neovide/neovide#1573]
I believe it would be great to have something similar to when nvim is running in Putty on Windows. In there, `Shift` + `mouse select` just selects and copies to clipboard the text on the screen, so that you can copy the status line, or edited text with line numbers, for example. Maybe in both plain text and rich text formats, with syntax highlighting.

In a similar way, `Shift` + `middle mouse click` could emulate pasting by taking text on the clipboard and emulating key presses for each symbol.

The point here is to have a plan B for the normal copy/paste that would work no matter what, and maybe also have some special use cases of its own.

The key combinations are just an example. Maybe there could be no hard-coded key combinations at all, just registered functions in nvim server that would send special commands back to the neovide client to enable the above. Those functions could be mapped to some key combos in nvim, if the user needs them.

**Comment by khjorth** [comment]
> The point here is to have a plan B for the normal copy/paste

Given the peculiarities of `vim.g.clipboard` and multiple UI clients, this has potential to be a good plan A.

> that would work no matter what

That's a tall order :)

> Maybe there could be no hard-coded key combinations at all, just registered functions in nvim server that would send special commands back to the neovide client to enable the above.

I think you will have a higher chance of achieving the goal by avoiding RPCs from Neovim to Neovide. If you go down this road, you're likely to end up with something that resembles the current solution in important ways and is thus liable to suffer from the same kind of issues. For instance, one major issue with the existing clipboard was that sometimes the RPCs from Neovim never reached Neovide, due to a missing or wrong channel ID.

Neovide already has support for manipulating the clipboard with RPCs from Neovim, although the one for getting the clipboard contents does some line-ending manipulation and also returns a paste mode for some reason. Just bear in mind that they are currently undocumented and not part of the external interface as far as I know, so they could change in future releases:

```lua
local lines = {'line1', 'line2'}

// Broadcast
vim.rpcrequest(0, 'neovide.set_clipboard', lines)
vim.rpcrequest(0, 'neovide.get_clipboard')

// Using the undocumented vim.g.neovide_channel_id variable
vim.rpcrequest(vim.g.neovide_channel_id, 'neovide.set_clipboard', lines)
vim.rpcrequest(vim.g.neovide_channel_id, 'neovide.get_clipboard')
```

Do they work no matter what? Certainly not - they have been broken in the past by the channel ID issue mentioned above and also by an integration issue with the clipboard library used in Neovide. The latter would affect any solution that interacts with the clipboard. A solution where Neovide is requested from Neovim to do the manipulation by emulating keystrokes would be affected by both.

But I think there's real promise in a more Putty like approach, where the clipboard manipulation is initiated in Neovide. This would avoid the hassle with `vim.g.clipboard` and RPCs from Neovim and should also provide sane behavior if someone were to attach multiple UIs to a single session - the latter is difficult with the current approach. Someone with knowledge of the editor part of the code base would have to weigh in on whether the mouse selection method you describe is feasible to implement.

**Comment by MultisampledNight** [comment]
Likely requires #1478 to be implemented first, and I thought I've seen an issue like this already but I can't seem to find it.

**Comment by j-xella** [comment]
> > that would work no matter what
> 
> That's a tall order :)

:) I meant a mechanism that would work as long as neovide is running, no matter what state the underlying nvim is, and if clipboard integration is set up properly or not ...

> For instance, one major issue with the existing clipboard was that sometimes the RPCs from Neovim never reached Neovide, due to a missing or wrong channel ID.

As for the wrong channel id (and all consequent problems), which may happen if multiple instances of neovide are connected to the same nvim server aat the same time, has it been considered to update clipboard integration functions (and whatever else can use the channel id) not just on connection, but every time, say, neovide window gains focus and user starts typing, or something similar?

**Comment by khjorth** [comment]
> :) I meant a mechanism that would work as long as neovide is running, no matter what state the underlying nvim is, and if clipboard integration is set up properly or not ...

Fair enough, that would be a substantial improvement over the current state of affairs.

> As for the wrong channel id (and all consequent problems), which may happen if multiple instances of neovide are connected to the same nvim server at the same time, has it been considered to update clipboard integration functions (and whatever else can use the channel id) not just on connection, but every time, say, neovide window gains focus and user starts typing, or something similar?

I thought about it briefly while tinkering with the clipboard, but haven't actually tried it. My main concern is that the approach seems likely to cause race conditions, that in turn could result in bugs that are difficult to diagnose.

The update-on-user-input approach could be easily implemented if Neovim provided a way to obtain the most recently used UI channel, but I can't find any such functionality in the current API. We would essentially add that feature through Neovide in a roundabout way. This gives me pause, because I would expect such functionality to be present in Neovim already if it were safe to use it. This is speculation though; it could be that no one else has needed it yet, or that it's a pain to implement.

The same concern applies to updating the state on focus gain or something similar. Here I would also be worried about half-broken window managers triggering the event even though they shouldn't, or not triggering the event even though they should.

For these reasons, I think the local approach where Neovide initiates the clipboard action is superior. I'd be happy to be proven wrong though.

More broadly speaking, I wonder if there is some feasible way to add UI client awareness to Neovim's clipboard handling. It's a major problem that you have to set `vim.g.clipboard` to a single provider that can somehow deal with multiple different UIs, potentially concurrently, when there's no established clipboard protocol that UIs can follow. I've searched the Neovim repo for information about this but haven't found anything so far.

**Comment by ZeChArtiahSaher** [comment]
I honestly think clipboard should work something like it does in tmux when you customize it with a mouse. In tmux visual mode is like a mode when you just copy literally any text you want. Cuz I find myself having to idk break my brain of simply having to copy text out of the nvim command line prompt

**Comment by fredizzimo** [comment]
This is quite tricky to implement, or at least a bit time consuming.

We need to both be able to track and select things, and that includes multiple lines. The selection should probably not cross multiple windows neovim Windows, but if it does, then it gets even more tricky. since we store the lines of each window separately.

Then we need to parse those selection coordinates and transform it into text, currently the stored format is not really text, but text with styling and other information.

We also need to implement a client-side mode for selecting and moving, and when that mode is active process the key inputs on the client side and don't pass them onto neovim. Then we need some kind of mapping to activate the mode and send a RPC request to activate it. And probably some kind of configuration options for the keybindings when the mode is active.

The actual copy and pasting to the clipboard is trivial since we have a library for it. But there are likely some things I have missed here, that additionally needs to be done.

@ZeChArtiahSaher, but for copying from the command-line, the easiest way is to type `q:` (not `:q`)  in normal mode or `<C-F>` in command mode, and it will open up a window with your command history, where you can copy from normally (you can also execute commands from there). See http://neovim.io/doc/user/cmdline.html#cmdwin

Pasting can be done with `<C-R>+`, or again using the same buffer, paste and edit the command there and execute it.

Copying the messages output is a bit more tricky, but you can use `:let @+ = execute('messages')` to copy it to your clipboard (or any other register, just replace `+`)

**Comment by aqilc** [comment]
I hope this can get some love this year! I would love to get a feature like this :)
